tests/functional/tests/smoke.py

```diff
@@ -0,0 +1,132 @@
+import logging
+import os
+import unittest
+from pathlib import Path
+from subprocess import STDOUT, CalledProcessError, check_call, check_output
+
+import zaza
+
+log = logging.getLogger(__name__)
+
+
+class SmokeTest(unittest.TestCase):
+    """COU smoke functional tests."""
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        zaza.get_or_create_libjuju_thread()
+        cls.create_local_share_folder()
+        cls.model_name = zaza.model.get_juju_model()
+        cls.install_package()
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        cls.remove_snap_package()
+        zaza.clean_up_libjuju_thread()
+
+    def create_local_share_folder() -> None:
+        """Create the .local/share/ folder if does not exist."""
+        Path(f"/home/{os.getenv('USER')}/.local/share/").mkdir(parents=True, exist_ok=True)
+
+    @classmethod
+    def install_package(cls) -> None:
+        """Install cou package."""
+        cou_test_snap = os.environ.get("TEST_SNAP")
+        if cou_test_snap:
+            log.info(f"Installing {cou_test_snap}")
+            assert Path(cou_test_snap).is_file()
+            # install the snap
+            assert check_call(f"sudo snap install --dangerous {cou_test_snap}".split()) == 0
+            # connect interfaces
+            interfaces_to_connect = [
+                "juju-client-observe",
+                "dot-local-share-cou",
+                "ssh-public-keys",
+            ]
+            for interface in interfaces_to_connect:
+                assert (
+                    check_call(
+                        f"sudo snap connect charmed-openstack-upgrader:{interface} snapd".split()
+                    )
+                    == 0
+                )
+            # make the cou alias
+            assert check_call("sudo snap alias charmed-openstack-upgrader.cou cou".split()) == 0
+
+            # check that the executable path exists
+            assert Path("/snap/bin/cou").exists()
+            cls.exc_path = "/snap/bin/cou"
+
+        else:
+            # functest already installs cou as python package, but we install again
+            # for better developer experience
+            log.warning("using cou as python package")
+            assert check_call(f"python3 -m pip install {os.getenv('PYTHONPATH')}".split()) == 0
+            cls.exc_path = "cou"
```

Comment:
I think that this can be ambitious, because if I installed cou with snap the `cou` will represent by this or not? Can we use something else like `./tox/func/bin/cou`?

Reply:
cou as python package is already installed via tox and it has preference because it's on the same virtual env. If you install the snap and call just by `cou` the snap won't be called. However, this is more declarative and I think we can make something as you mentioned

Reply:
I think which executable path `cou` refers to is depending on which `cou` is installed first. In functional tests, python package is installed first through tox, `cou` will always point to the python executable. I agree that the path should be more explicit if possible, but in case we can't get the actual path in code (e.g. `./tox/func` vs `./tox/func-smoke` created by different tox target ), I think using just `cou` with some explanatory comments is fine